runtime/caches.go

```diff
@@ -105,82 +110,31 @@ func (c *connectionCache) evict(ctx context.Context, instanceID, driver, dsn str
 	return ok
 }
 
-type catalogCache struct {
-	cache map[string]*catalog.Service
-	lock  sync.Mutex
-}
-
-func newCatalogCache() *catalogCache {
-	return &catalogCache{
-		cache: make(map[string]*catalog.Service),
-	}
+type migrationMetaCache struct {
+	cache *lru.Cache
 }
 
-func (c *catalogCache) get(ctx context.Context, rt *Runtime, instID string) (*catalog.Service, error) {
-	// TODO 1: opening a driver shouldn't take too long but we should still have an instance specific lock
-	// TODO 2: This is a cache on a cache, which may lead to undefined behavior
-	// TODO 3: Use LRU and not a map
-
-	c.lock.Lock()
-	defer c.lock.Unlock()
-
-	key := instID
-
-	service, ok := c.cache[key]
-	if ok {
-		return service, nil
-	}
-
-	registry, _ := rt.metastore.RegistryStore()
-	inst, err := registry.FindInstance(ctx, instID)
-	if err != nil {
-		return nil, err
-	}
-
-	olapConn, err := rt.connCache.get(ctx, instID, inst.OLAPDriver, inst.OLAPDSN)
+func newMigrationMetaCache(size int) *migrationMetaCache {
+	cache, err := lru.New(size)
 	if err != nil {
-		return nil, err
+		panic(err)
 	}
-	olap, _ := olapConn.OLAPStore()
-
-	var catalogStore drivers.CatalogStore
-	if inst.EmbedCatalog {
-		conn, err := rt.connCache.get(ctx, inst.ID, inst.OLAPDriver, inst.OLAPDSN)
-		if err != nil {
-			return nil, err
-		}
-
-		store, ok := conn.CatalogStore()
-		if !ok {
-			return nil, fmt.Errorf("instance cannot embed catalog")
-		}
 
-		catalogStore = store
-	} else {
-		store, ok := rt.metastore.CatalogStore()
-		if !ok {
-			return nil, fmt.Errorf("metastore cannot serve as catalog")
-		}
-		catalogStore = store
-	}
+	return &migrationMetaCache{cache: cache}
+}
 
-	repoConn, err := rt.connCache.get(ctx, instID, inst.RepoDriver, inst.RepoDSN)
-	if err != nil {
-		return nil, err
+func (c *migrationMetaCache) get(instID string) *catalog.MigrationMeta {
+	if val, ok := c.cache.Get(instID); ok {
+		return val.(*catalog.MigrationMeta)
 	}
-	repoStore, _ := repoConn.RepoStore()
 
-	service = catalog.NewService(catalogStore, repoStore, olap, registry, instID, rt.logger)
-	c.cache[key] = service
-	return service, nil
+	meta := catalog.NewMigrationMeta()
+	c.cache.Add(instID, meta)
+	return meta
 }
```

Comment:
There's a race condition here – imagine concurrent calls to `get`, both fail the `if` check, both will create a new `MigrationMeta` and return separate objects

Reply:
This is why `simplelru.LRU` and a mutex seemed better for `connectionCache` than `lru.Cache`

Reply:
I was under the assumption that `only` one migration for one instance will run in all circumstances even with the redesigned reconcile. Is it not the case ? 

The problem with using `simplelru.LRU` and a mutex is that it unnecessarily blocks under instances as well. Although agreed our critical section here is not intensive and the overhead is going to be negligible but if we know that no two instances will ever run this in parallel its a nice trade-off to make even from the point of view of reuse of existing constructs.

> This is why simplelru.LRU and a mutex seemed better for connectionCache than lru.Cache

IMO this is okay for `migrationMetaCache` but certainly not okay for `connectionCache` if we expect to run thousands of instance on a single runtime host. The `connectionCache.Get` call is in the hot path of many APIs and having a single mutex for all instances may not be good for a multi tenant system.
One solution could be to find something similar to [guava Cache](https://guava.dev/releases/19.0/api/docs/com/google/common/cache/LoadingCache.html) or build it ourselves if nothing is available.

Reply:
On second thought, lets change it to `simplelru.LRU` and a mutex as you mentioned since the underlying map anyways blocks the access to all threads and we aren't doing any intensive operation after taking lock.